### PR TITLE
docs: Expand training customization examples

### LINKED ISSUE #4379: Remove or populate "Training customization"
Currently, this part of the documentation shows some possible customizations that applies to all trainers https://huggingface.co/docs/trl/main/en/customization

However, it only features a few examples. This sections would make sense if it gets populated with other customizations, or removed. This thread can be used to discussed additional customizations

### PR DESCRIPTION
**Pull request by behroozazarkhalili**
## Summary

This PR addresses issue #4379 by **expanding** the Training Customization documentation section with 5 new comprehensive examples, rather than removing it.

Resolves #4379

### New Examples Added (5):
1. **Custom Callbacks** - Shows how to add custom callbacks for logging, monitoring, or early stopping
2. **Custom Evaluation Metrics** - Demonstrates defining custom metrics to track during training
3. **Mixed Precision Training** - Explains bf16/fp16 usage for speed and memory optimization
4. **Gradient Accumulation** - Shows how to simulate larger batch sizes with limited GPU memory
5. **Custom Data Collator** - Demonstrates custom data preprocessing and padding strategies

### Documentation Improvements:
- Updated introduction for better clarity and consistency
- All examples follow the same pattern as existing examples
- All code examples verified against the codebase
- Proper imports and configuration options validated

## Statistics

- **Original examples**: 5
- **New examples**: 5
- **Total examples**: 10 (doubled!)
- **Lines added**: ~150

## Verification

✅ All imports verified against codebase  
✅ All config options verified in DPOConfig  
✅ DataCollatorForPreference import path corrected  
✅ Consistent code style with existing examples  
✅ Examples apply to most/all trainers as stated

## Test Plan

- [x] Verified all imports exist in the codebase
- [x] Validated config parameters against DPOConfig
- [x] Ensured consistent formatting with existing examples
- [x] Checked that examples follow DPOTrainer pattern as stated in intro

### CODE REVIEW

File: docs/source/customization.md [hunk 1]
```diff
@@ -117,3 +117,152 @@ When training large models, you should better handle the accelerator cache by it
...
+trainer = DPOTrainer(
+    model=model,
+    args=training_args,
+    train_dataset=dataset,
+    tokenizer=tokenizer,
+)
+trainer.train()
+```
+
+Note: Use `bf16=True` for Ampere GPUs (A100, RTX 30xx) or newer, and `fp16=True` for older GPUs.
```
Comment (qgallouedec) [inline docs/source/customization.md:117]: maybe it's worth making it clearer that bf16 is the default value in TRL

File: docs/source/customization.md [hunk 2]
```diff
@@ -117,3 +117,152 @@ When training large models, you should better handle the accelerator cache by it
...
+    gradient_accumulation_steps=8,
+)
+
+trainer = DPOTrainer(
+    model=model,
+    args=training_args,
+    train_dataset=dataset,
+    tokenizer=tokenizer,
+)
+trainer.train()
```
Comment (qgallouedec) [inline docs/source/customization.md]: this section would be more relevant in "reducing memory" section. It's more a trick than a a customization

**Comment by behroozazarkhalili** [comment]
I've addressed both review comments:

1. **Mixed precision section**: Added clarification that `bf16=True` is the default in TRL. Updated the example to show when/how to override defaults for older GPUs or to disable mixed precision.

2. **Gradient accumulation section**: Moved from customization guide to the reducing memory usage guide (`reducing_memory_usage.md`), as it's primarily a memory optimization technique. Expanded the examples to include DPO, SFT, and Reward trainers.

Ready for re-review!

**Review by sergiopaniego — COMMENTED** [review]
Thanks!! 
I'd aim on reducing the code snippets content, only including the relevant part. Otherwise it's complicated to understand what's been added. A good example of this is the subsection `Use the accelerator cache optimizer`, where it directly points to the added param. 
Btw the `optimize_device_cache` is no longer part of the codebase so that subsection can actually be removed :)

File: docs/source/customization.md [hunk 3]
```diff
@@ -1,6 +1,6 @@
 # Training customization
 
-TRL is designed with modularity in mind so that users are able to efficiently customize the training loop for their needs. Below are some examples on how you can apply and test different techniques.  Note: Although these examples use the DPOTrainer, the customization applies to most (if not all) trainers.
+TRL is designed with modularity in mind so that users are able to efficiently customize the training loop for their needs. Below are examples on how you can apply and test different techniques. Note: Although these examples use the [`DPOTrainer`], these customization methods apply to most (if not all) trainers in TRL.
```
Comment (sergiopaniego) [inline docs/source/customization.md]:
```suggestion
TRL is designed with modularity in mind so that users are able to efficiently customize the training loop for their needs. Below are examples on how you can apply and test different techniques. 

> [!NOTE]
> Although these examples use the [`DPOTrainer`], these customization methods apply to most (if not all) trainers in TRL.
```

File: docs/source/customization.md [hunk 4]
```diff
@@ -89,31 +61,79 @@ Since `trl` supports all keyword arguments when loading a model from `transforme
 Read more about 8-bit model loading in `transformers` [Load in 8bit or 4bit](https://huggingface.co/docs/transformers/en/peft#load-in-8bit-or-4bit).
```
Comment (sergiopaniego) [inline docs/source/customization.md]:
```suggestion
Read more about 8-bit model loading in `transformers` [Load in 8bit or 4bit](https://huggingface.co/docs/transformers/en/peft).
```

**Review by sergiopaniego — APPROVED** [review]
thanks!
